mask download url when there is a password

jinxiao opened the pull request:
# Pull Request

<!--
The text in these markdown comments is instructions that will not appear in the displayed pull request.
-->

mask download url when there is a password.

## Problem

<!--
What problem are you solving? Include issue numbers if it has been reported. 
Show the broken output if appropriate.
-->
When I use export N_NODE_MIRROR with username and password, I found that the password is a plaintext which cause our security concern.
## Solution

<!--
How did you solve the problem? 
Show the fixed output if appropriate.
-->

I introduce a new variable called masked_url, when called log or something want to output the url, can use this variable.

## ChangeLog

<!--
Optional. Suggest a line for adding to the CHANGELOG to summarise your change.
-->

shadowspawn commented:
I am open to this. There are quite a log of places an URL gets displayed. One place I am particularly aware of is in `show_diagnostics`, enough that I put in a warning `Note: some output may contain passwords. Redact before sharing.`.

I suggest move the masking code into a function, say `display_masked_url`, so it can be used in more places.

jinxiao commented:
> I am open to this. There are quite a log of places an URL gets displayed. One place I am particularly aware of is in `show_diagnostics`, enough that I put in a warning `Note: some output may contain passwords. Redact before sharing.`.
> 
> I suggest move the masking code into a function, say `display_masked_url`, so it can be used in more places.

Good idea, I have changed this. Could you please have a check?  I also changed the download URL in diagnostics to avoid any password leakage.

shadowspawn commented on bin/n:
```diff
@@ -793,15 +797,15 @@ install() {
 
   local url="$(tarball_url "$version")"
-  is_ok "${url}" || abort "download preflight failed for '$version' (${url})"
+  is_ok "${url}" || abort "download preflight failed for '$version' '$(display_masked_url ${url})'"
```
The password may have a space in it, so the url needs to be quoted to pass the expansion as a single argument:
```suggestion
  is_ok "${url}" || abort "download preflight failed for '$version' '$(display_masked_url "${url})"'"
```

(Applies to other uses too.)

shadowspawn commented on bin/n:
```diff
@@ -1494,8 +1498,8 @@ function show_diagnostics() {
-  echo "node mirror: ${N_NODE_MIRROR}"
-  echo "node downloads mirror: ${N_NODE_DOWNLOAD_MIRROR}"
+  echo "node mirror: $(display_masked_url ${N_NODE_MIRROR})"
+  echo "node downloads mirror: $(display_masked_url "${N_NODE_DOWNLOAD_MIRROR}")"
```
(This one has quotes already.) 👍

shadowspawn commented on bin/n:
```diff
@@ -793,15 +797,15 @@ install() {
 
   local url="$(tarball_url "$version")"
-  is_ok "${url}" || abort "download preflight failed for '$version' (${url})"
+  is_ok "${url}" || abort "download preflight failed for '$version' '$(display_masked_url "${url})"'"
```
```suggestion
  is_ok "${url}" || abort "download preflight failed for $version $(display_masked_url "${url}")"
```

shadowspawn commented on bin/n:
```diff
@@ -793,15 +797,15 @@ install() {
   cd "${dir}" || abort "Failed to cd to ${dir}"
 
-  log fetch "$url"
+  log fetch "$(display_masked_url ${url})"
```
```suggestion
  log fetch "$(display_masked_url "${url}")"
```

shadowspawn commented on bin/n:
```diff
@@ -1241,7 +1245,7 @@ display_remote_index() {
   if [[ "${PIPESTATUS[0]}" -ne 0 ]]; then
     # Reminder: abort will only exit subshell, but consistent error display
-    abort "failed to download version index (${index_url})"
+    abort "failed to download version index $(display_masked_url ${index_url})"
```
```suggestion
    abort "failed to download version index $(display_masked_url "${index_url}")"
```

shadowspawn commented on bin/n:
```diff
@@ -1494,8 +1498,8 @@ function show_diagnostics() {
   printf "\nn\n"
-  echo "node mirror: ${N_NODE_MIRROR}"
-  echo "node downloads mirror: ${N_NODE_DOWNLOAD_MIRROR}"
+  echo "node mirror: $(display_masked_url ${N_NODE_MIRROR})"
```
```suggestion
  echo "node mirror: $(display_masked_url "${N_NODE_MIRROR}")"
```

shadowspawn commented:
The quotes are a bit tricky with the nesting!

shadowspawn commented:
(Closed by mistake.)

shadowspawn commented on bin/n:
```diff
@@ -793,15 +797,15 @@ install() {
 
   local url="$(tarball_url "$version")"
-  is_ok "${url}" || abort "download preflight failed for '$version' (${url})"
+  is_ok "${url}" || abort "download preflight failed for $version $(display_masked_url "${url}")"
```
Preserving original style now I realise the single quotes and brackets appear in the output.
```suggestion
  is_ok "${url}" || abort "download preflight failed for '$version' ($(display_masked_url "${url}"))"
```

shadowspawn commented on bin/n:
```diff
@@ -1241,7 +1245,7 @@ display_remote_index() {
   if [[ "${PIPESTATUS[0]}" -ne 0 ]]; then
     # Reminder: abort will only exit subshell, but consistent error display
-    abort "failed to download version index (${index_url})"
+    abort "failed to download version index $(display_masked_url "${index_url}")"
```
```suggestion
    abort "failed to download version index ($(display_masked_url "${index_url}"))"
```

shadowspawn commented:
I misunderstood a couple of the uses, so added suggestions to preserve the original style.

shadowspawn commented:
I was worried about spaces in password, but now I remember username and password need to be url-encoded for robustness anyway: https://github.com/tj/n/blob/master/docs/proxy-server.md

However, adding the quotes around the variable expansions fixes warnings from `shellcheck` so I would have wanted the changes anyway.

shadowspawn commented:
Thanks @jinxiao

shadowspawn commented:
Released in v10.1.0
